client/mifarehost.c

```diff
@@ -840,16 +883,62 @@ int mfTraceDecode(uint8_t *data_src, int len, bool wantSaveToEmlFile) {
 				lfsr_rollback_word(revstate, uid ^ nt, 0);
 
 				crypto1_get_lfsr(revstate, &lfsr);
-				printf("key> %x%x\n", (unsigned int)((lfsr & 0xFFFFFFFF00000000) >> 32), (unsigned int)(lfsr & 0xFFFFFFFF));
+				ui64Key = lfsr;
+				printf("key> %x%x Prng:%s\n", 
+					(unsigned int)((lfsr & 0xFFFFFFFF00000000) >> 32), (unsigned int)(lfsr & 0xFFFFFFFF), 
+					validate_prng_nonce(nt) ? "WEAK": "HARDEND");
 				AddLogUint64(logHexFileName, "key> ", lfsr);
 			} else {
-				printf("key> nested not implemented!\n");
-				at_enc = bytes_to_num(data, 4);
+				if (validate_prng_nonce(nt)) {
+					struct Crypto1State *pcs;
+					pcs = crypto1_create(ui64Key);
+					uint32_t nt1 = crypto1_word(pcs, nt_enc ^ uid, 1) ^ nt_enc;
+					uint32_t ar = prng_successor(nt1, 64);
+					uint32_t at = prng_successor(nt1, 96);
+					printf("key> nested auth uid: %08x nt: %08x nt_parity: %s ar: %08x at: %08x\n", uid, nt1, printBitsPar(&nt_enc_par, 4), ar, at);
+					uint32_t nr1 = crypto1_word(pcs, nr_enc, 1) ^ nr_enc;
+					uint32_t ar1 = crypto1_word(pcs, 0, 0) ^ ar_enc;
+					uint32_t at1 = crypto1_word(pcs, 0, 0) ^ at_enc;
+					printf("key> the same key test. nr1: %08x ar1: %08x at1: %08x \n", nr1, ar1, at1);
+
+					if (NTParityCheck(nt1))
+						printf("key> the same key test OK. key=%x%x\n", (unsigned int)((ui64Key & 0xFFFFFFFF00000000) >> 32), (unsigned int)(ui64Key & 0xFFFFFFFF));
+					else
+						printf("key> the same key test. check nt parity error.\n");
+					
+					uint32_t ntx = prng_successor(nt, 90);
+					for (int i = 0; i < 16383; i++) {
+						ntx = prng_successor(ntx, 1);
+						if (NTParityCheck(ntx)){
+							printf("key> nt candidate=%08x nonce distance=%d\n", ntx, nonce_distance(nt, ntx));
+							break;
+						}						
+					}
+					
+					nt = ntx;
+					ks2 = ar_enc ^ prng_successor(ntx, 64);
+					ks3 = at_enc ^ prng_successor(ntx, 96);
+
+					// decode key
+					revstate = lfsr_recovery64(ks2, ks3);

```

Comment:
`revstate` is a pointer to list of possible keys, not a single key.

Reply:
Need `free(revstate)` somewhere.

Reply:
free done. 
added hint about possible key. I cant realize how i can check keys. in trace decoding we can just print probable key or list of several probable keys. all the keys in this list will generate ks2 and ks3(

Reply:
that is the extra tests from mf_nonce_brute is for,   basically it tries the candiate keys and decrypt the command afterwards..  if it matches one of IS14a commands and the crc is correct,   that key is considered to be valid/found.